irohad/maintenance/metrics.cpp

```diff
@@ -9,88 +9,128 @@
+  cvcbs.log_message = civet_no_log;
+  cvcbs.log_access = civet_no_log;
+
+  // create an http server running on addr:port
+  exposer_ = std::make_shared<Exposer>(listen_addr_port_,
+                                       /*num_threads*/ 2,
+                                       &cvcbs);
+
+  // ask the exposer_ to scrape the registry_ on incoming HTTP requests
+  exposer_->RegisterCollectable(registry_, "/metrics");
+
+  auto &block_height_gauge = BuildGauge()
+                                 .Name("blocks_height")
+                                 .Help("Total number of blocks in chain")
+                                 .Register(*registry_);
+  auto &block_height = block_height_gauge.Add({});
+  block_height.Set(storage_->getBlockQuery()->getTopBlockHeight());
+
+  auto &peers_number_gauge =
+      BuildGauge()
+          .Name("peers_number")
+          .Help("Total number peers to send transactions and request proposals")
+          .Register(*registry_);
+  auto &number_of_peers = peers_number_gauge.Add({});
+  number_of_peers.Set(storage_->getWsvQuery()->getPeers()->size());
+
+  auto &domains_number_gauge = BuildGauge()
+                                   .Name("number_of_domains")
+                                   .Help("Total number of domains in WSV")
+                                   .Register(*registry_);
+  auto &domains_number = domains_number_gauge.Add({});
+  domains_number.Set(storage_->getWsvQuery()->countDomains().assumeValue());
+
+  auto &total_number_of_transactions_gauge =
+      BuildGauge()
+          .Name("total_number_of_transactions")
+          .Help("Total number of transactions in blockchain")
+          .Register(*registry_);
+  auto &total_number_of_transactions =
+      total_number_of_transactions_gauge.Add({});
+  total_number_of_transactions.Set(
+      storage_->getWsvQuery()->countTransactions().assumeValue());
+
+  auto &number_of_signatures_in_last_block_gauge =
+      BuildGauge()
+          .Name("number_of_signatures_in_last_block")
+          .Help("Number of signatures in last block")
+          .Register(*registry_);
+  auto &number_of_signatures_in_last_block =
+      number_of_signatures_in_last_block_gauge.Add({});
+  auto ptopblock =
+      storage_->getBlockQuery()
+          ->getBlock(storage_->getBlockQuery()->getTopBlockHeight())
+          .assumeValue();
+  number_of_signatures_in_last_block.Set(boost::size(ptopblock->signatures()));
+
+  block_subscriber_ =
+      SubscriberCreator<bool, BlockPtr>::template create<EventTypes::kOnBlock>(
+          SubscriptionEngineHandlers::kMetrics,
+          [&, wregistry = std::weak_ptr<Registry>(registry_)](auto &,
+                                                              BlockPtr pblock) {
+            // Metrics values are stored inside and owned by registry,
+            // capture them by reference is legal.
+            std::shared_ptr<Registry> registry{wregistry};  // throw if expired
```

Comment:
Can avoid throw with `if (auto maybe_registry = registry.lock())` though

Reply:
This should never happen! expired pointer is critical logic error. Because in class Metrics subscribtion is dessruted before registry.